src/localFunction.ts

```diff
@@ -0,0 +1,225 @@
+import * as request from "request";
+
+import { encodeFirestoreValue } from "./firestore/encodeFirestoreValue";
+import * as utils from "./utils";
+import { EmulatedTriggerDefinition } from "./emulator/functionsEmulatorShared";
+import { FunctionsEmulatorShell } from "./emulator/functionsEmulatorShell";
+import { AuthMode } from "./emulator/events/types";
+
+/**
+ * @class
+ * @this LocalFunction
+ * @param {object} trigger
+ * @param {object=} urls
+ * @param {object=} controller
+ */
+export default class LocalFunctionClass {
+  private url?: string;
+
+  constructor(
+    private trigger: EmulatedTriggerDefinition,
+    urls: Record<string, string>,
+    private controller: FunctionsEmulatorShell
+  ) {
+    this.url = urls[trigger.id];
+  }
+
+  private substituteParams(resource: string, params: Record<string, string>) {
+    const wildcardRegex = new RegExp("{[^/{}]*}", "g");
+    return resource.replace(wildcardRegex, (wildcard: string) => {
+      const wildcardNoBraces = wildcard.slice(1, -1); // .slice removes '{' and '}' from wildcard
+      const sub = params?.[wildcardNoBraces];
+      return sub || wildcardNoBraces + utils.randomInt(1, 9);
+    });
+  }
+  private constructCallableFunc(data: any, opts: any) {
```

Comment:
How tricky would it be to have stronger types here? I think data could be `string | object | undefined`, and opts can be `opts?: { instanceIdToken?: string } = {}` (or you can make a shared type that covers all the different `opts` in this file